### ISSUE TITLE
MT-32 timing incorrect in Legend Entertainment games

### ISSUE THREAD

**Issue by johnnovak** [dosbox-staging/dosbox-staging#1271]
### Are you using the latest Dosbox-Staging Version?

- [X] I have checked releases and am using the latest release.

### Different version than latest?

latest master (06d86622)

### What Operating System are you using?

Windows 10

### Have you checked that no other similar issue already exists?

- [X] I have searched and not found similar issues.

### A clear and concise description of what the bug is.

Just a heads up, MIDI timing is _severely_ messed up in the latest master (06d86622) compared to 0.77.1 in *certain* games. This is very likely related to possible regressions the MPU-401 emulation code.

Note this happens only in _certain games_, most notably **all** Legend Entertainment games that use MPU-401 Intelligent Mode. You can't just test it with any game, in most games the issue probably doesn't manifest at all (e.g. Sierra games I tested are not affected).

I've tested with **Eric the Unready** and **Spellcaster 101**, and it's 99% certain all other Legend Entertainment games are affected too (see [this Vogons thread](https://www.vogons.org/viewtopic.php?f=33&t=46178&start=20) for a detailed explanation).

You can also compare the Eric the Unready intro music to a pristine recording of the actual MIDI:
https://www.youtube.com/watch?v=er4oXmfjn8s&list=PLTg6d_6DGTy80cQ7sJiDFIS6t2V3A5yZx&index=2

### Steps to reproduce the behaviour.

1. Start Eric the Unready using 0.77.1 and let the music play for about a minute. Note that apart from some global tempo fluctuations, the music is mostly fine (the fluctuations are due to the MPU-401 emulation not being 100% accurate, as explained [in this Vogons thread](https://www.vogons.org/viewtopic.php?f=33&t=46178&start=20) -- but that's a separate issue).
2. Repeat with latest from master (I used 06d86622). Note that the timing of different instrument tracks will progressively drift away from each other. After 30 secs it's already a complete unlisteneable mess.
3. Repeat the same experiment with Spellcaster 101. The issue is perhaps even more audible just after about 15 seconds into the music.
4. For Eric the Unready, you can use this recording as a reference: https://www.youtube.com/watch?v=er4oXmfjn8s&list=PLTg6d_6DGTy80cQ7sJiDFIS6t2V3A5yZx&index=2

### Your configuration

```ini
[sdl]
fullscreen = true

[midi]
mididevice = mt32

[mt32]
model = cm32l
romdir = d:\Emulators\DOS\ROMs\Roland\

[autoexec]
mount C "C"
imgmount D "CD\Eric the Unready (1997)(Legend).iso" -t iso
c:
legend
```

### Provide a Log

```text
I don't think the log will help in this case at all, but here it is:

dosbox-staging version 0.77.1-1826d
---
CONFIG: Loading primary file C:\Users\johnn\AppData\Local\DOSBox\dosbox-staging.conf
CONFIG: Loading additional file dosbox.conf
CONFIG: Unknown option maxresolution
MAIN: Sized window to 971x728 with stretched pixels
OPENGL: Pixel buffer object extension: available
SDL: Mouse will move seamlessly without being captured and middle-click will uncapture the mouse.
MEMORY: Base address: 000001A40E0AD040
MEMORY: Using 4096 DOS memory pages (16 MiB)
MIXER: Negotiated 2-channel 44100-Hz audio in 1024-frame blocks
MT32: Initialized CM-32L/LAPC-I Control v1.02 from d:\Emulators\DOS\ROMs\Roland\
MIXER: MT32 channel operating at 44100 Hz without resampling
MIDI: Opened device: mt32
MIXER: FM channel operating at 44100 Hz without resampling
MIXER: SB channel operating at 22050 Hz and upsampling to the output rate
MIXER: SPKR channel operating at 18939 Hz and upsampling to the output rate
MIXER: DISNEY channel operating at 10000 Hz and upsampling to the output rate
MAPPER: Found no joysticks
MAPPER: Loaded default key bindings
MOUNT: Path 'C' found
SDL: captured the mouse
IMGMOUNT: Path 'CD\Eric the Unready (1997)(Legend).iso' found
MIXER: CDAUDIO channel operating at 44100 Hz without resampling
MAIN: Draw resolution: 640x400, pixel aspect ratio: 1.20
MAIN: Draw resolution: 640x480, pixel aspect ratio: 1.00
MT32: LCD-Message:  ERIC THE UNREADY
SB16: DSP was reset
SB16: Speaker-output has been toggled on
MAIN: Draw resolution: 320x200, double-width, double-height, pixel aspect ratio: 1.20
SDL: released the mouse
```

### Code of Conduct & Contributing Guidelines

- [X] Yes, I agree.

**Comment by johnnovak** [comment]
Extra info: just tested `v0.78.0-alpha` and the regression is not present in that branch (behaves exactly like 0.77.1).

**Comment by kcgen** [comment]
Thanks for the testing @johnnovak.

I do have these two games, but unfortunately I see that both are running in MT-32 modes.  Do you know how to configure them to use general MIDI?

**Comment by johnnovak** [comment]
Perhaps I was not entirely clear in my use terminology; when I wrote MIDI I meant "MT-32 MIDI", not General Midi.

These are my MIDI settings:

```
[midi]
mididevice = mt32

[mt32]
model = cm32l
```

And this is how I'm starting the game:

```
ERIC XGA XMOUSE BLASTER 7 220 MT32 2 330
```

**Comment by johnnovak** [comment]
At the risk of overexplaining: the root cause is most likely changes the MPU-401 emulation is DOSBox and/or the PIC emulation. Most Legend Entertainment games use MPU-401 Intelligent Mode for their MT-32 MIDI music, and they interface with the MPU-401 in non-standard ways (as far as I understand it, they don't send data to the MPU in its own interrupt, but in the timer interrupt a bit later, plus they reprogram the PIC so timer interrupts happen at a much faster rate. See [this thread](https://www.vogons.org/viewtopic.php?f=33&t=46178) for a detailed description (plus note they're solving an entirely different issue in that thread, that's nothing to do with this regression, but it could be relevant)).

**Comment by arrowgent** [comment]
hmmm... i can test deathgate and callahan's crosstime saloon
linux x86_64
dosbox-staging 0.78.1 WxH branch

deathgate, ive played this twice with general midi from beginning to end.
testing mpu401 & mt32 in the sound options
you have to redo the installation to change sound options (or manually edit legend.ini)
i dont seem to have any issues with mt32, cm32l, or fluidsynth soundfont
each option sounds different and behaves fluidly

LEGEND.INI
```
MUSIC=MPU401.MDI GM
SOUND=SBPRO.DIG
```


callahan's ive never played, only installed and tested the intro sequences.
switching between mt32 & fluidsynth in dosbox settings works fine,  maybe one or two songs dont seem to play.  im not sure if thats a game mechanic limitation or not. just at the intro with the keyboard player requesting songs

MDI.INI
```
DEVICE      General MIDI (Roland MPU-401 interface or 100% compatible)
DRIVER      MPU401.MDI
IO_ADDR     330h
IRQ         -1
DMA_8_BIT   -1
DMA_16_BIT  -1
```

**Comment by kcgen** [comment]
@arrowgent re: **0.78.1 WxH branch**; can you please try with the latest sources?

We do have a known audio timing-related glitch (https://github.com/dosbox-staging/dosbox-staging/issues/1257), which @kklobe will be tackling.

**Comment by johnnovak** [comment]
> i dont have the MPU TSR that vogons is talking about, and they seem to note its only for earlier (or specific) games from legend.
> 
> and reading back on it again, the midi files themselves play fine outside of the game engine. they are suggesting the game engine used by legend is having an issue with BPM/tempo, not dosbox itself.
> 
> of course i also dont have "real hardware" to compare tempo speed bugs against. it would be trivial to compare 5% speed difference with acual recordings. testing that the music audio plays is the best i can do

I'm afraid the discussion is going a bit sideways here... You have tested this with a different branch and two different games than I!

Without going into detail territory, the regression is easy to demonstrate:

- test the Eric the Unready and Spellcasting 101 intro music on MT-32 with 0.77.1 or 0.78.0-alpha - **music plays fine**
- then retest with 06d8662 (or latest master, most likely it's still unfixed) - **the timing of the music is completely messed up** (but sure, it plays *something... 😄) 

I will upload some recordings later today to make 100% sure we're talking about the same regression. Again, let's not go into the technical details on what Legend did or how in their game engine, I'm simply taking about a regression that happened since 0.77.1.

I also regret that I brought up that TSR from the Vogons forum because it just muddies the waters... let's forget about it and just focus on the regression now 😄 

For completeness sake, I will also re-test with Death Gate and Callahan today and report back.

**Comment by kcgen** [comment]
Thanks for getting this back on the rails @johnnovak. 
Will be checking this out tomorrow.

**Comment by johnnovak** [comment]
Okay, here's my results:

@arrowgent Death Gate and Callahan's Crosstime Saloon are *not* affected, indeed; the music plays perfectly fine on all DOSBox versions that I tried. Your theory was right, only earlier Legend Entertainment games seem to be affected. I tried Spellcasting 201 and 301 as well, and the regression can be observed on those early titles too.

So this is a list of Legend Entertainment games that are 100% confirmed to be affected by the MT-32/MPU-401 regression:

- Spellcasting 101
- Spellcasting 201
- Spellcasting 301
- Eric the Unready (SVGA CD-ROM version)

There's probably more but this should be plenty to reproduce the bug.

**_It is important to test this with these 4 games as the vast majority of other MT-32 titles are NOT affected by the regression!_** (e.g. none of the Sierra games I tried were affected)

----

And as promised, here are some audio recordings that demonstrate the issue. Note how the timing between the various instruments is completely out of sync in the "bad" versions -- as if the "band members" were seriously drunk 😁 This is especially apparent in the Spellcaster 101 intro music.

**0.77.1 & v0.78-alpha**

[Spellcasting 101 - good](https://drive.google.com/file/d/1IV5oVq6lWNKGQ7YOWOJcjpBu8-mSjORO/view?usp=sharing)
[Eric the Unready - good](https://drive.google.com/file/d/14kBtAuWL3MdoUH3nr1D6rn5Tfw1VgIBa/view?usp=sharing)

**latest master** (c22a0e5973)

[Spellcasting 101 - bad](https://drive.google.com/file/d/18vH3yr-yIw-tecC81IqeqaEDOvOsdvpv/view?usp=sharing)
[Eric the Unready - bad](https://drive.google.com/file/d/1df134rp5GZgaqzJ5H6Zy8kD7d3UvIpU2/view?usp=sharing)
